Provide elephpant data in JSON

colinodell opened the pull request:
This bundle creates a `data.json` file with all the elephpants.  This can then be used in other elephpant-related projects (like https://github.com/matthewtrask/elephpant-traders/issues/36)

Requires #70

colinodell commented:
Here's the file it creates: https://gist.github.com/colinodell/46320f6fbc1c66cedc246ee307330cf7

colinodell commented:
Rebased onto `master` (to pull in the `composer.json` changes)

philipsharp commented:
I've finally had some time to look over this implementation.

There are a few fields in the output that I think would need to be cleaned up. The "e-*" ID format was a quick hack to support in-page anchors, and I don't want to use it like this. This is also exposing a lot of the Sculpin tricks like using tags for the color and categories for the variation that should be hidden to a consumer of the data.

I don't know if it's possible in this implementation to reference the site configuration, which I would use to add more detail to the output like the variation names.

Overall this is a lot of code to produce JSON output. I realize that much of that is because it's extending Sculpin the right way with a Symfony bundle, but it does create an ongoing maintenance requirement. The code doesn't appear to be re-usable, even if we just wanted to create a second data feed.

I'm considering a simpler alternative to use Twig to generate the JSON, similar to https://github.com/thatpodcast/thatpodcast.io/blob/master/source/episodes.json . I'd prefer a way to execute raw PHP inside a page template but without that this approach feels easier to maintain.

colinodell commented:
I appreciate the feedback!

I made some changes to simplify the implementation and resolve some of your concerns.  I know that still doesn't meet your ideal solution of using Twig, but perhaps it might be good enough to hold us over for now?

> The "e-*" ID format was a quick hack to support in-page anchors, and I don't want to use it like this.

Sure, I have removed this and replaced it with the Markdown filename (minus `.md`).  This allows consumers to reference a unique identifier which won't change even if the name/description/etc. does.

> This is also exposing a lot of the Sculpin tricks like using tags for the color and categories for the variation that should be hidden to a consumer of the data.

I've renamed those fields so the consumer doesn't know about the ~dirty hacks~ Sculpin tricks being used :wink: 

> I don't know if it's possible in this implementation to reference the site configuration, which I would use to add more detail to the output like the variation names.

That should be possible - let me know what you had in mind.

> Overall this is a lot of code to produce JSON output. I realize that much of that is because it's extending Sculpin the right way with a Symfony bundle, but it does create an ongoing maintenance requirement.

Yeah, most of that was due to following Symfony conventions.  However, Symfony is flexible enough that we don't really need all the boilerplate.  I was therefore able to reduce this down to 3 files:

 - The bundle file (which also defines the service and event hook)
 - A reusable class for adding dynamic JSON feeds
 - The elephpant data converter

(If we didn't care about reusability or stateless services we could combine the latter 2 classes)

> The code doesn't appear to be re-usable, even if we just wanted to create a second data feed.

The new `DynamicJsonSource` I created is reusable :)  We can use this to add additional data feeds as needed.

> I'm considering a simpler alternative to use Twig to generate the JSON

I totally understand if you'd still prefer that approach.

colinodell commented:
Updated JSON output: https://gist.github.com/colinodell/46320f6fbc1c66cedc246ee307330cf7

colinodell commented on src/JsonBundle/DynamicJsonSource.php:
```diff
@@ -0,0 +1,36 @@
+/**
+ * A dynamically-generated JSON feed of data
+ */
+class DynamicJsonSource extends AbstractSource
```
This is the reusable class for creating JSON data feeds.

ramondelafuente commented:
I like the json-api approach. What I think would be a good addition to the raw data is the available sizes for each elephpant, and in some cases the 'batch' that was produced (only blue and pink probably have multiple batches, all the others where only produced once).

colinodell commented on src/JsonBundle/JsonBuilder.php:
```diff
@@ -0,0 +1,47 @@
+            'reverse' => @$data['reverse'],
+            'photos' => @$data['photos'],
+            'color' => @$data['tags'][0],
+            'variation' => @$data['categories'][0],
```
Some elephpants are missing some of these fields, which is why I'm suppressing any warnings here.  (I wanted to keep this code as short as possible).

If you're building the site with PHP 7, we could use the null coalescing operator (`??`).

We could also wrap the entire thing in `array_filter()` to remove empty/null fields if desired.

philipsharp commented:
I have a better understanding of how this works after the rewrite.

As for the site configuration, they're loaded from https://github.com/philipsharp/afieldguidetoelephpants/blob/master/app/config/sculpin_site.yml and exposed in the templates as `site.*`.  It primarily allows reusing data between different "populations" of a single variation (like different Zend elephpants). Things like the variation name and latin name:
https://github.com/philipsharp/afieldguidetoelephpants/blob/master/source/_sections/3-variations.html#L23-L24

FYI the markdown filename is *mostly* unchanging. There were a lot of changes in the first few months while I caught up on existing elephpants. I don't think I've renamed any in the past years since now it's pretty clear when a new elephpant appears. It's a better identifier than the "e-*" hack, which I used just because the filename wasn't available in the twig templates.

colinodell commented:
I've updated my code to pull those Latin and common names - see https://gist.github.com/colinodell/46320f6fbc1c66cedc246ee307330cf7/revisions

philipsharp commented:
I had to add the PSR-4 autoload information to composer.json. I remember seeing it before; I think this got lost when you rebased.

Otherwise I made a few tweaks after merging. The file is now available at: http://afieldguidetoelephpants.net/data/all.json

Thanks for your help! Sorry it took so long to get merged.

matthewtrask commented:
This looks fantastic. Im going to work on this at PHP World when Im hanging around! Thank you both for all of your hard work!
